tests/NuGet.IndexingTests/VersionResultTests.cs

```diff
@@ -0,0 +1,253 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Lucene.Net.Support;
+using NuGet.Indexing;
+using NuGet.Versioning;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Xunit;
+
+namespace NuGet.IndexingTests
+{
+    public class VersionResultTests
+    {
+        private readonly Random randomizer = new Random(1000000);
+
+        [Theory]
+        [MemberData(nameof(VersionStringSets))]
+        public void LegacyResultHasNoSemVer2(string[] versionSet, int expectedNumSemVer1)
+        {
+            var result = MakeVersionResult(versionSet, randomizeListing: true);
+
+            var legacyResult = result.LegacyVersionDetails;
+            Assert.Equal(expectedNumSemVer1, legacyResult.Count());
+            foreach (var details in legacyResult)
+            {
+                Assert.True(!details.IsSemVer2);
+            }
+        }
+        
+        [Theory]
+        [MemberData(nameof(VersionStringSets))]
+        public void SemVer2ResultContainsAllVersion(string[] versionSet, int expectedNumSemVer1)
+        {
+            var result = MakeVersionResult(versionSet, randomizeListing: true);
+
+            var semVer2Result = result.SemVer2VersionDetails;
+            Assert.True(semVer2Result.Count() >= expectedNumSemVer1);
+            Assert.True(semVer2Result.Count() == versionSet.Length);
+        }
+
+        [Theory]
+        [MemberData(nameof(VersionStringSets))]
+        public void StableResultContainsNoPrerelVersion(string[] versionSet, int expectedNumSemVer1)
+        {
+            var result = MakeVersionResult(versionSet, randomizeListing: true);
+
+            var semVer2StableResult = result.GetStableVersions(onlyListed: false, includeSemVer2: true);
+            var semVer2StableListedResult = result.GetStableVersions(onlyListed: true, includeSemVer2: true);
+            var semVer1StableResult = result.GetStableVersions(onlyListed: false, includeSemVer2: false);
+            var semVer1StableListedResult = result.GetStableVersions(onlyListed: true, includeSemVer2: false);
+
+            var semVer2StableCount = 0;
+            var semVer2StableListedCount = 0;
+            var semVer1StableCount = 0;
+            var semVer1StableListedCount = 0;
+
+            foreach (var details in semVer2StableResult)
```

Comment:
nit:

You can condense this code a lot. It won't be as efficient as the way you wrote it, but it'll be easier to read.

```
Assert.True(semVer2StableResult.All(v => !v.IsPrerelease);
Assert.True(semVer2StableListedResult.All(v => !v.IsPrerelease);
Assert.True(semVer1StableResult.All(v => !v.IsPrerelease);
Assert.True(semVer1StableListedResult.All(v => !v.IsPrerelease);

Assert.True(semVer2StableResult.Count() >= semVer1StableResult.Count());
Assert.True(semVer2StableResult.Count() >= semVer2StableListedResult.Count());
Assert.True(semVer1StableResult.Count() >= semVer1StableListedResult.Count());
Assert.True(semVer2StableListedResult.Count() >= semVer1StableListedResult.Count());
```